### PR TITLE
AT_FDCWD support.

### PR DESCRIPTION
**Pull request by sunfishcode**
This forward-ports #54 to current main.

### CODE REVIEW

**Comment by sunfishcode** [comment]
A brief description of what this is:

>    AT_FDCWD is a special constant in POSIX that can be passed to *at
    functions to indicate the current working directory. Since the
    current working directory is emulated in wasi libc, add emulated
    AT_FDCWD support as well.

File: Makefile [hunk 1]
```diff
@@ -188,6 +188,10 @@ CFLAGS = $(WASM_CFLAGS) --target=$(TARGET_TRIPLE)
 # WebAssembly floating-point match doesn't trap.
 # TODO: Add -fno-signaling-nans when the compiler supports it.
 CFLAGS += -fno-trapping-math
+# We should have proper declarations for everything.
+CFLAGS += \
+    -Werror=missing-declarations \
+    -Werror=implicit-function-declaration \
```
Comment (sbc100) [inline Makefile:194]: Can we instead use `-Werror` and disabled the warngins that are not fixable or not part in code that we maintain?

File: libc-bottom-half/cloudlibc/src/libc/fcntl/openat.c [hunk 2]
```diff
@@ -11,156 +11,61 @@
 #include <fcntl.h>
 #include <string.h>
 
-#ifdef __wasilibc_unmodified_upstream // fstat
-static_assert(O_APPEND == __WASI_FDFLAG_APPEND, "Value mismatch");
-static_assert(O_DSYNC == __WASI_FDFLAG_DSYNC, "Value mismatch");
-static_assert(O_NONBLOCK == __WASI_FDFLAG_NONBLOCK, "Value mismatch");
-static_assert(O_RSYNC == __WASI_FDFLAG_RSYNC, "Value mismatch");
-static_assert(O_SYNC == __WASI_FDFLAG_SYNC, "Value mismatch");
-#else
```
Comment (sbc100) [inline libc-bottom-half/cloudlibc/src/libc/fcntl/openat.c:20]: There removals of the ifdef blocks seem unrelated?    Could they land separately?

File: Makefile (hunk 1 above)
Comment (sbc100) [reply Makefile:194]: Also maybe this should be its own change.

**Comment by sunfishcode** [comment]
The warnings and ifdef changes are now landed as separate chnages, and this is now rebased on top of them.

**Review by sbc100 — APPROVED** [review]
